### ISSUE TITLE
Memory leak when subscribing to observables from XAML in a data template

### ISSUE THREAD

**Issue by ShadowsInRain** [AvaloniaUI/Avalonia#5872]
### Bug description

Subscribing to an observable from XAML (using a caret `^` syntax, e.g. `IsVisible="{Binding Foo^}"`) in a data template causes a memory leak: the binding recipient remains in the memory after it's long gone from the visual tree.

Likely related to #3527

### Reproduction

Bind to a long-lived observable in a data template.

Minimal example, based on the default template: [repo link](https://github.com/ShadowsInRain/MWE20210504-AvaloniaUI) (most relevant stuff is in the `MainWindow`). Dummy controls are used to track a leak.

To confirm a leak, I've used the managed memory profiler in VS 2019.
![image](https://user-images.githubusercontent.com/1629574/117021611-eea7fa00-ad10-11eb-9540-0b777f9981af.png)

### Expected behavior

All bindings made in XAML should be auto-disposed when control is detached. That seems to be true when using normal bindings (e.g. `IsVisible="{Binding $parent[UserControl].Foo}"` — included in the example).

### Desktop

* Windows 10 Pro (v2004)
* Avalonia 0.10.3

### Commentary

In my 100%-reactive app this issue causes colossal memory leaks: just switching a page bumps the memory gauge by 50-70 Mb. It stacks nicely with (up to) 350 Mb from the graphics bug (the one that causes memory issues when resizing a window).

I do not know if the issues is caused by a data template or a binding, but having the bug down the tree causes a huge chunk of the tree to leak, likely through a parent reference.

**Comment by ShadowsInRain** [comment]
After updating to `0.10.6` the leak is still present, #5987 did not help. Repro repo is up-to-date. No new screenshots, reference count didn't change.

**Comment by ShadowsInRain** [comment]
I think the leak may be happening somewhere in the pre-compiled (a)XAML. Although the compiler internals are confusing. Can someone point me in the general direction?

**Comment by grokys** [comment]
Taking a look now @ShadowsInRain

**Comment by grokys** [comment]
Yeah this is a difficult one. I've been unable to write a unit test so far that reproduces the problem, even. Seems to be related to the fact that an observable is producing a collection via the stream operator which creates data templates which themselves have observables subscribed via the stream operator.

**Comment by unofficialdev** [comment]
I'm having a similar problem. i try to rewrite 1 wpf app with avalonia, wpf app only use ~50mb memory, but avalonia app go up to ~400mb

**Comment by sdijoseph** [comment]
I believe I just ran into this issue on 11.3.0

**Comment by aguahombre** [comment]
Using conditional classes with caret ^ syntax also leaks memory; e.g.
```Classes.active="{Binding Item^.IsActive}"```
Avalonia 11.3.0

**Comment by grokys** [comment]
The brave and scary new world of LLMs has allowed me (OK not me - Claude) to write a failing test for this, five years later. Hopefully it can make a fix too.
